enhancements/operator-sdk-builds-catalogs.md

```diff
@@ -0,0 +1,179 @@
+or at least would like to test their bundles collectively as a catalog.
+`operator-sdk` already facilitates building and testing bundles, so it should
+facilitate building catalogs from these bundles.
+
+## Motivation
+
+Canonically, `opm index add` is how a catalog is created and bundles are added to a catalog.
+In fact, it is the only entrypoint into the world of managing catalogs, likely for adoption reasons and
+because these processes are largely opaque and complicated, even though the spec is not.
+While `opm` is intended for "power users" (read: those who need to manipulate catalog images),
+certain subcommands like `index` and its related libraries can be used fairly easily by anyone
+`operator-sdk` should expose functionality of `opm` in a limited and sanely-defaulted way
+to enable building catalog images.
+
+### Goals
+
+- Expose `opm index add` functionality to `operator-sdk` users for all supported Operator types.
+- Use existing libraries/`opm` itself to expose the desired functionality.
+- Increase understanding of the Operator packaging lifecycle.
+
+### Non-Goals
+
+- Re-implement any part of the catalog management process.
+- Add any catalog-related code or files to `operator-sdk` projects, ex. `index.Dockerfile`.
+- Add any other functionality other than build a catalog locally (`opm index add`).
+
+## Proposal
+
+### User Stories
+
+#### Build a catalog using `opm`
+
+As an Operator developer, I want to easily build a catalog from my Operator's bundles
+without having to know the details of _how_ it is built.
+
+#### Understand what a catalog is
+
+As an Operator developer, I should understand what a catalog is and why I would consider
+building one. I also should understand how bundling and building a catalog ties into shipping my Operator.
+
+### Implementation Details/Notes/Constraints
+
+There are two ways of adding `opm index add` functionality to `operator-sdk`:
+1. Create a `catalog-build` Makefile rule that calls `opm index add`.
+1. Add a `operator-sdk catalog build` subcommand (discussed in the [alternatives](#subcommand-approach) section).
+
+The Makefile approach is by far the best for a few reasons:
+- Users have full control over which `opm` version to use.
+- Users have full control over defaults for their project, including image tags.
+- Users get experience using `opm`, but only the relevant part(s).
+  - They can easily add more `opm` targets if they want.
+
+The Makefile additions would look like:
+```make
+VERSION ?= 0.1.0
+IMAGE_TAG_BASE ?= quay.io/example/my-operator
+BUNDLE_IMG ?= $(IMAGE_TAG_BASE)-bundle:$(VERSION)
+
+bundle-push:
+	$(MAKE) docker-push IMG=$(BUNDLE_IMG)
+
+OPM = ./bin/opm
+opm:
+  # Download `opm` from github.com/operator-framework/operator-registry/releases
+
+BUNDLE_IMGS ?= $(BUNDLE_IMG)
+CATALOG_IMG ?= $(IMAGE_TAG_BASE)-catalog:v$(VERSION)
+catalog-build: opm
+	$(OPM) index add --mode semver --tag $(CATALOG_IMG) --bundles $(BUNDLE_IMGS)
```

Comment:
Question:
If I want to use an existing catalog image I assume I could just do:

`make catalog-build CATALOG_IMG=quay.io/me/my-image-with-deps:dev`

And then we will add our bundle to that existing image.

Reply:
There's actually another flag `--from-index` you'd use to build from an existing index image. Since the base image doesn't actually exist anywhere (`opm index add` builds it for you), I've opted not to include this flag. Perhaps a Makefile comment about this flag would be useful.

Reply:
I'm not sure I understand this comment @estroz , `--from-index` is for cumulative builds. It allows you to add a single bundle to an existing index by pulling the old index down, pulling the content onto local disk, appending to it, and then building a new index without having to pull all the bundles every time you just want to update your existing set. Are we saying that the usecase for the sdk expects their users to keep historical record of all previous bundle images locally?

Reply:
> Are we saying that the usecase for the sdk expects their users to keep historical record of all previous bundle images locally?

I don't think "keep a historical record of all previous bundle images locally" is a requirement. Hence, it's better to allow users to add a new bundle image (and potentially its dependencies from other Operators) to an existing catalog.

Reply:
>`--from-index` is for cumulative builds

That's what I intend users to do.

>Are we saying that the usecase for the sdk expects their users to keep historical record of all previous bundle images locally

Definitely not, that's what git is for.

Reply:
I'm just trying to point out that `--from-index` is a pretty core piece of functionality that we expect all of our opm users to be using aside from the first time they publish an index. It seems odd to differ that fundamentally from how most folks are using the tool now. Or maybe I'm just totally misreading this?

Reply:
Its more a problem of defaults than differentiating behavior. There is no default image to set `--from-index` to when a project is created such that I can run `make catalog-build` without `opm index add` failing. The user has to build an index image containing one bundle first, then add `--from-index` to the `catalog-build` recipe one they're ready to add another bundle to the existing index. That's why I suggest we add a comment about this.